File: src/shared/domain/value_object/contact/url_value_object.py
from dataclasses import dataclass
from urllib.parse import urlparse

from ..primitive import StringValueObject
from ..invalid_argument_exception import InvalidArgumentException


@dataclass(frozen=True)
class UrlValueObject(StringValueObject):
    """A value object that represents a URL."""

    _value: str

    def __post_init__(self) -> None:
        try:
            result = urlparse(self._value)
            if not all([result.scheme, result.netloc]):
                raise ValueError("Invalid URL format")
        except Exception as e:
            raise InvalidArgumentException(
                field="url", value=self._value, message="Invalid URL format"
            ) from e

    @property
    def domain(self) -> str:
        """Get the domain part of the URL."""
        return urlparse(self._value).netloc

    @property
    def scheme(self) -> str:
        """Get the scheme (protocol) of the URL."""
        return urlparse(self._value).scheme

    def to_primitive(self) -> str:
        return self._value

File: hexy/typescript/src/shared/domain/value-objects/contact/url-value-object.ts
import { StringValueObject } from '../primitives/string-value-object'
import { InvalidValueObjectError } from '../errors/invalid-value-object-error'

/**
 * Represents a validated URL
 */
export class UrlValueObject extends StringValueObject {
	constructor(value: string) {
		try {
			const url = new URL(value)
			super(url.toString())
		} catch {
			throw new InvalidValueObjectError('Invalid URL format')
		}
	}

	protected validate(value: string): boolean {
		try {
			new URL(value)
			return true
		} catch {
			return false
		}
	}

	/**
	 * Gets the URL domain
	 */
	get domain(): string {
		return new URL(this.value).hostname
	}

	/**
	 * Gets the URL protocol
	 */
	get protocol(): string {
		return new URL(this.value).protocol.replace(':', '')
	}
}
